docs/eeps/draft-new-plugin-tests-with-junit5.md

```diff
@@ -0,0 +1,214 @@
+
+Split tests to simple unit tests and "integrated" Embulk plugin tests
+----------------------------------------------------------------------
+
+Simple unit tests, which just test the behavior of a single class, do not require the special class loading structure. The class loading structure is required only when testing its behavior as an Embulk plugin.
+
+It is, however, uneasy to modify the class loading structure during a set of test execution. It would be reasonable to split tests into two parts:
+
+1. Simple unit tests
+2. "Integrated" Embulk plugin tests
+
+The simple unit tests are common Java tests. On the other hand, the "integrated" Embulk plugin tests run with the special class loading structure. In terms of `embulk-junit4`, tests that required `@Rule EmbulkTestRuntime runtime` would be "integrated" Embulk plugin tests. Others would be simple unit tests.
+
+Collaborate with Gradle and the Gradle plugin
+----------------------------------------------
+
+Ordinary Embulk plugins are built with Gradle and [the `org.embulk.embulk-plugins` Gradle plugin](https://plugins.gradle.org/plugin/org.embulk.embulk-plugins). A natural approach to distinguish the simple unit tests and the "integrated" Embulk plugin tests is to run them as two different Gradle test tasks. For example :
+
+```
+./gradlew test  # Run simple unit tests.
+./gradlew embulkPluginTest  # Run the "integrated" Embulk plugin tests.
+```
+
+At the same time, Gradle with the Gradle plugin is the good point to tweak the class loading structure. Indeed, it is difficult to tweak the class loading structure once the test execution has started.
+
+The `org.embulk.embulk-plugins` Gradle plugin would define a new `EmbulkPluginTest` task type that extends the standard [`Test`](https://docs.gradle.org/8.5/javadoc/org/gradle/api/tasks/testing/Test.html) type, and registers its default task `embulkPluginTest`. It would also define default [`Configuration`](https://docs.gradle.org/8.5/dsl/org.gradle.api.artifacts.Configuration.html)s and [`SourceSet`](https://docs.gradle.org/8.5/dsl/org.gradle.api.tasks.SourceSet.html)s. The `EmbulkPluginTest` task would configure the class loading structure with dependencies for testing.
+
+The `build.gradle` would be like below.
+
+```gradle
+plugins {
+    id "java"
+    id "org.embulk.embulk-plugins" version "..."
+}
+
+// ...
+
+dependencies {
+    compileOnly "org.embulk:embulk-spi:..."
+
+    implementation "org.embulk:embulk-util-config:..."
+    // ...
+
+    // Simple unit tests.
+    testImplementation "org.junit.jupiter:junit-jupiter-api:..."
+    testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine:..."
+
+    // "Integrated" Embulk plugin tests.
+    embulkPluginTestImplementation "org.junit.jupiter:junit-jupiter-api:..."
+
```

Comment:
Just question, Does "Integrated" Embulk plugin test use `org.junit.jupyter`? not `org.embulk..`?

Reply:
That's another point I still wonder. However, without jupiter-api, we'll have to implement so many things, including all `assert`s and else. I expect we'll want at least jupiter-api...